client/java/src/main/java/com/linecorp/centraldogma/client/DefaultWatcher.java

```diff
@@ -28,108 +28,74 @@
-            nextDelayMillis = Math.min(
-                    saturatedMultiply(MIN_INTERVAL_MILLIS, Math.pow(2.0, numAttemptsSoFar - 1)),
-                    MAX_INTERVAL_MILLIS);
-        }
-
-        final long minJitter = (long) (nextDelayMillis * (1 - JITTER_RATE));
-        final long maxJitter = (long) (nextDelayMillis * (1 + JITTER_RATE));
-        final long bound = maxJitter - minJitter + 1;
-        final long millis = random(bound);
-        return Math.max(0, saturatedAdd(minJitter, millis));
-    }
-
-    private static long saturatedMultiply(long left, double right) {
-        final double result = left * right;
-        return result >= Long.MAX_VALUE ? Long.MAX_VALUE : (long) result;
-    }
-
-    private static long random(long bound) {
-        assert bound > 0;
-        final long mask = bound - 1;
-        final Random random = ThreadLocalRandom.current();
-        long result = random.nextLong();
-
-        if ((bound & mask) == 0L) {
-            // power of two
-            result &= mask;
-        } else { // reject over-represented candidates
-            for (long u = result >>> 1; u + mask - (result = u % bound) < 0L; u = random.nextLong() >>> 1) {
-                continue;
-            }
-        }
+final class DefaultWatcher<T> implements Watcher<T> {
 
-        return result;
-    }
+    private static final Logger logger = LoggerFactory.getLogger(DefaultWatcher.class);
 
     private enum State {
         INIT,
         STARTED,
         STOPPED
     }
 
-    private final CentralDogma client;
     private final ScheduledExecutorService watchScheduler;
     private final String projectName;
     private final String repositoryName;
     private final String pathPattern;
+    private final Function<Revision, CompletableFuture<Latest<T>>> watchFunction;
+    private final boolean errorOnEntryNotFound;
+    private final WatcherOptions watcherOptions;
 
-    private final List<Map.Entry<BiConsumer<? super Revision, ? super T>, Executor>> updateListeners;
-    private final AtomicReference<State> state;
-    private final CompletableFuture<Latest<T>> initialValueFuture;
+    private final List<Map.Entry<BiConsumer<? super Revision, ? super T>, Executor>> updateListeners =
+            new CopyOnWriteArrayList<>();
+    private final AtomicReference<State> state = new AtomicReference<>(State.INIT);
+    private final CompletableFuture<Latest<T>> initialValueFuture = new CompletableFuture<>();
 
+    @Nullable
     private volatile Latest<T> latest;
+    @Nullable
     private volatile ScheduledFuture<?> currentScheduleFuture;
+    @Nullable
     private volatile CompletableFuture<?> currentWatchFuture;
 
-    protected AbstractWatcher(CentralDogma client, ScheduledExecutorService watchScheduler,
-                              String projectName, String repositoryName, String pathPattern) {
-        this.client = requireNonNull(client, "client");
+    DefaultWatcher(ScheduledExecutorService watchScheduler,
```

Comment:
That's a good suggestion. 👍 
So basically the file watcher and files watcher are created in `WatchRequest` and `WatchFilesRequest` respectively, but you want to move the creating logic into the static methods of this `DefaultWatcher`.

Well either way is fine to me. 😄  But I slightly prefer to leave it as it is because `DefaultWatcher` is pretty complex now so I want to keep it as simple as possible to reduce the cognitive load. In contrast, `WatchRequest` class is simple so there's not so much distraction.

Like I said, either way is fine so please let me know if you want it strongly, then I'll move. 😄

Reply:
I also think `DefaultWatcher` is pretty complex.
So it is not bad to create new instance in `WatchRequest` and `WatchFilesRequest`.
I was just worried that `watchFunction` looks like key function to `DefaultWatcher`  :)

Reply:
Based on the comment, I revived `FileWatcher` and `FilesWatcher` to distinguish that. PTAL. 😄